### PR TITLE
refactor depart_at and arrive_by and solve some issues

### LINKED ISSUE #462: Heading for through locations unreliable/sensitive
(2016-12-09T23:25:11Z) @ericrwolfe : originally created valhalla/thor#518

Something seems off with heading as a parameter for through/break locations.

As an example, here's a path with 3 locations. The heading between the 2nd and 3rd points is ~174°, but simply changing the heading between 174 and 175 results in two completely different route results.

"heading": 174: Correct path, distance 0.856

```
POST http://valhalla.mapzen.com/route
{
  "costing":"pedestrian",
  "locations":[{
    "lon":-122.4225686207969,
    "heading":81,
    "lat":37.75679855287286,
    "type":"break"
  },
  {
    "lat":37.758636,
    "lon":-122.419076,
    "heading": 174,
    "type":"through"
  },
  {
    "lat":37.75543499999999,
    "lon":-122.41877,
    "type":"break"
  }]
}
```

"heading": 175: Incorrect path, distance 1.213km with a confusing & costly u-turn maneuver

```
POST http://valhalla.mapzen.com/route
{
  "costing":"pedestrian",
  "locations":[{
    "lon":-122.4225686207969,
    "heading":81,
    "lat":37.75679855287286,
    "type":"break"
  },
  {
    "lat":37.758636,
    "lon":-122.419076,
    "heading": 175,
    "type":"through"
  },
  {
    "lat":37.75543499999999,
    "lon":-122.41877,
    "type":"break"
  }]
}
```

(2016-12-13T18:53:17Z) @dnesbitt61 : Thanks for the report. I have been able to duplicate and will investigate more soon.

(2016-12-13T20:02:58Z) @dnesbitt61 : Some more info:

This demonstrates a through location that is basically at a node. Likely needs changes in both loki and thor (with possible additions to PathLocation)

loki - for heading =  174 (finds 4 edges - opposing inbound and outbound I think for 20th St) but for 175 it gives back all 8 edges (inbound/outbound). Probably if it snaps to a node it should not filter by heading. But to best handle, it needs to return all but identify the edge that most closely matches the heading.

thor: If a destination is a though location AND at a node it should allow all edges except the one most closely aligned to the heading for the destination of the first segment. This will allow the best path to the node (but not along the path that requires a Uturn). When updating the origin locations for the next segment it should use the edge marked (by loki) as most closely aligned.

Also, line 353 in route_action.cc should be:   for(const auto& e : destination.edges)  (not the origin edges since how can the last edge in the path be the origin edge?)

We probably won't be able to get to this until after the New Year. As a near term workaround make sure through locations are not at nodes.

### PR DESCRIPTION
**Pull request by kevinkreiser**
this fixes a small bug where turn costs at through locations would not be counted. this only happened when a through location was at a node. it also fixes #462 which was originally the main point of this work.

basically when a through location is used, we force the algorithm to use the same edge twice (the one that had the through on it) regardless of whether it was at a node or not. this required a bit of change in the way the depart_at and arrive_by stuff worked and how the algorithms excluded certain edges at nodes depending on whether it was a origin or destination.

do note that the original issue #462 was talking about pedestrian routes. this change will still include a u-turn for pedestrian routes who snap to nodes and use a heading to limit the route to leaving that location in a specific direction. the reason is because pedestrian routes do not penalize u-turns. to enforce that we'd need to allow a costing option to penalize u-turns for pedestrian routes or change the idea of how a node snapped through location with a heading should work (ie. force the path to come into the location in the direction that it leaves #problematic)

ive also added some test requests so we can do some before and after.

### CODE REVIEW

File: src/loki/search.cc [hunk 1]
```diff
@@ -189,6 +189,19 @@ PathLocation correlate_node(GraphReader& reader, const Location& location, const
 
   //start where we are and crawl from there
   crawl(found_node, true);
+
+
+  //if it was a through location with a heading its pretty confusing.
```
Comment (kevinkreiser) [inline src/loki/search.cc]: this is the secret sauce behind fixing #462

**Comment by dnesbitt61** [comment]
:shipit:

**Comment by gknisely** [comment]
did you RAD the transit tests

**Comment by kevinkreiser** [comment]
@gknisely i did not RAD transit. but ill do that, the trick will be telling failures of data vs failures of the algorithm :smile:.  i did run both arrive and depart routes while testing. in fact im going to add some regular arrive by routes just to have them..

**Comment by gknisely** [comment]
@dgearhart and I know the transit tests pretty well....if they fail ask away.

**Pull request merged**